gnovm/pkg/gnolang/preprocess.go

```diff
@@ -737,6 +747,88 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 				cx := evalConst(store, last, n)
 				return cx, TRANS_CONTINUE
 
+			// TRANS_LEAVE -----------------------
+			case *FuncLitExpr:
+				// We need to closure-call if any variables
+				// used were declared in a for-loop, and this
+				// is the outer-most function within the loop.
+				// NOTE: if the variable is defined as a
+				// function parameter of the closure, then it
+				// is shadowed and not relevant whether or not
+				// the closure is immediately called.
+
+				// Step 1: check if this func lit is in a loop,
+				// not embedded within another func lit.
+				loopDepth, inLoop := isBlockNodeInLoop(store, n)
+				if !inLoop {
+					return n, TRANS_CONTINUE
+				}
+
+				// Step 2: gather all extern names declared in for-loops.
+				leNames := []Name{}
+				leTypes := []Type{}
+				lePaths := []ValuePath{}
+				for _, name := range n.GetExternNames() {
+					path := n.GetExternPathForName(store, name)
+					if path.Type != VPBlock {
+						continue // not a for-loop block path
+					}
+					if int(path.Depth) <= loopDepth+1 {
+						// it is within the loop, carry on...
+					} else {
+						// it is outside the closest loop ancestor,
+						// check to see if it is declared in outer loop.
+						container := n.GetBlockNodeForPath(store, path)
+						if _, ok := isBlockNodeInLoop(store, container); !ok {
+							continue // not declared in any loop.
+						}
+					}
+					typ := n.GetStaticTypeOfAt(store, path)
+					// Append loop extern name and related.
+					leNames = append(leNames, name)
+					lePaths = append(lePaths, path)
+					leTypes = append(leTypes, typ)
+				}
+				if len(leNames) == 0 {
+					// nothing to transform
+					return n, TRANS_CONTINUE
+				}
+				panic("!!! ITS HAPPENING (insert ron paul jazz hands) !!!")
```

Comment:
NOTE: I haven't actually tested the running of this logic, yet.

Reply:
"TestSearchEfficiency" is triggered by this but the tests have been passing because the closure is called synchronously. good sign that it is picked up by this panic, and probably a good one to focus on first.